File: hard/030-substring-with-concatenation-of-all-words/solution.py
from typing import List
from collections import Counter, defaultdict


class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        """
        30. Substring with Concatenation of All Words
        Time: O(n * L)
        Space: O(m * L)
        """
        n = len(s)
        L = len(words[0])
        m = len(words)
        total = L * m
        if total > n:
            return []

        need = Counter(words)
        result = []

        # Windows only align to L distinct offsets; scan each track separately.
        for offset in range(L):
            window = defaultdict(int)
            left = offset
            count = 0

            for right in range(offset, n - L + 1, L):
                word = s[right:right + L]

                if word not in need:
                    # No valid window can span an unknown word: restart past it.
                    window.clear()
                    count = 0
                    left = right + L
                    continue

                window[word] += 1
                count += 1

                # Too many copies of this word: shrink from the left.
                while window[word] > need[word]:
                    window[s[left:left + L]] -= 1
                    left += L
                    count -= 1

                if count == m:
                    result.append(left)
                    # Slide one word forward to look for the next match.
                    window[s[left:left + L]] -= 1
                    left += L
                    count -= 1

        return result

File: hard/030-substring-with-concatenation-of-all-words/solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 30. Substring with Concatenation of All Words
 * Time: O(n * L)
 * Space: O(m * L)
 */
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        int n = s.length();
        int L = words[0].length();
        int m = words.length;
        if ((long) L * m > n) {
            return result;
        }

        Map<String, Integer> need = new HashMap<>();
        for (String w : words) {
            need.merge(w, 1, Integer::sum);
        }

        // Windows only align to L distinct offsets; scan each track separately.
        for (int offset = 0; offset < L; offset++) {
            Map<String, Integer> window = new HashMap<>();
            int left = offset;
            int count = 0;

            for (int right = offset; right + L <= n; right += L) {
                String word = s.substring(right, right + L);

                if (!need.containsKey(word)) {
                    // No valid window can span an unknown word: restart past it.
                    window.clear();
                    count = 0;
                    left = right + L;
                    continue;
                }

                window.merge(word, 1, Integer::sum);
                count++;

                // Too many copies of this word: shrink from the left.
                while (window.get(word) > need.get(word)) {
                    String lw = s.substring(left, left + L);
                    window.merge(lw, -1, Integer::sum);
                    left += L;
                    count--;
                }

                if (count == m) {
                    result.add(left);
                    // Slide one word forward to look for the next match.
                    String lw = s.substring(left, left + L);
                    window.merge(lw, -1, Integer::sum);
                    left += L;
                    count--;
                }
            }
        }

        return result;
    }
}
